File: dashlane-project/Dashlane/src/main/java/com/dashlane/loaders/InstalledAppAndPopularWebsiteLoader.java
package com.dashlane.loaders;

import android.content.Context;

import com.dashlane.ui.util.DataLoaderForUi;
import com.dashlane.util.domain.PopularWebsiteUtils;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

import androidx.fragment.app.Fragment;

public class InstalledAppAndPopularWebsiteLoader extends DataLoaderForUi<List<String>> {

    public interface Listener {
        void onLoadFinished(@Nullable List<String> result);
    }

    private final Fragment mFragment;
    private final Listener mListener;

    public InstalledAppAndPopularWebsiteLoader(@NotNull Fragment fragment, Listener listener) {
        super(fragment);
        mFragment = fragment;
        mListener = listener;
    }

    @Override
    protected List<String> loadData() {
        Context context = mFragment.getContext();
        if (context == null) return new ArrayList<>();
        return PopularWebsiteUtils.getPopularWebsites(context);
    }

    @Override
    protected void onPostExecute(@Nullable List<String> result) {
        mListener.onLoadFinished(result);
    }
}

File: dashlane-project/Dashlane/src/main/java/com/dashlane/loaders/InstalledAppAndPopularWebsiteLoader.kt
package com.dashlane.loaders

import androidx.fragment.app.Fragment
import com.dashlane.ext.application.KnownApplicationProvider
import com.dashlane.ui.util.DataLoaderForUi
import com.dashlane.util.domain.PopularWebsiteUtils.getPopularWebsites

class InstalledAppAndPopularWebsiteLoader(
    private val fragment: Fragment,
    private val listener: Listener,
    private val knownApplicationProvider: KnownApplicationProvider
) : DataLoaderForUi<List<String?>?>(fragment) {
    interface Listener {
        fun onLoadFinished(result: List<String>?)
    }

    override fun loadData(): List<String> {
        val context = fragment.context ?: return ArrayList()
        return getPopularWebsites(context, knownApplicationProvider)
    }

    override fun onPostExecute(t: List<String?>?) {
        listener.onLoadFinished(t?.filterNotNull())
    }
}